eos-updater-flatpak-installer/eos-updater-flatpak-installer.c

```diff
@@ -0,0 +1,262 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright © 2017 Endless Mobile, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Authors:
+ *  - Sam Spilsbury <sam@endlessm.com>
+ */
+
+#include <stdlib.h>
+#include <glib.h>
+#include <locale.h>
+#include <flatpak.h>
+#include <ostree.h>
+#include <libeos-updater-util/util.h>
+#include <eos-updater-types.h>
+
+typedef struct _FlatpakToDeploy {
+  gchar *repo;
+  gchar *app_id;
+  FlatpakRefKind ref_kind;
+} FlatpakToDeploy;
+
+static FlatpakToDeploy *
+flatpak_to_deploy_new (const gchar    *repo,
+                       const gchar    *app_id,
+                       FlatpakRefKind ref_kind)
+{
+  FlatpakToDeploy *to_deploy = g_new0 (FlatpakToDeploy, 1);
+
+  to_deploy->repo = g_strdup (repo);
+  to_deploy->app_id = g_strdup (app_id);
+  to_deploy->ref_kind = ref_kind;
+
+  return to_deploy;
+}
+
+static void
+flatpak_to_deploy_free (FlatpakToDeploy *to_deploy)
+{
+  g_free (to_deploy->repo);
+  g_free (to_deploy->app_id);
+  g_free (to_deploy);
+}
+
+/* Sue me */
+static gboolean
+parse_flatpak_ref_kind (const gchar *string, FlatpakRefKind *kind)
+{
+  g_return_val_if_fail (kind != NULL, FALSE);
+
+  if (g_strcmp0 (string, "app") == 0)
+    {
+      *kind = FLATPAK_REF_KIND_APP;
+      return TRUE;
+    }
+
+  if (g_strcmp0 (string, "runtime") == 0)
+    {
+      *kind = FLATPAK_REF_KIND_RUNTIME;
+      return TRUE;
+    }
+
+  return FALSE;
+}
+
+static void
+malformed_flatpak_to_deploy_file_error (GError **error,
+                                        const gchar *entry)
+{
+  g_set_error (error,
+               EOS_UPDATER_ERROR,
+               EOS_UPDATER_ERROR_MALFORMED_FLATPAK_ENTRIES,
+               "Malformed entry %s in flatpak auto-deploy file "
+               "please consult your vendor",
+               entry);
+}
+
+static GPtrArray *
+parse_flatpaks_to_deploy_from_contents (const gchar *contents,
```

Comment:
I’d rather we didn’t have yet another file format and parser in the code. Perhaps we could just have one file per app? For example, something like `$remote_name/$flatpak_ref`, where the `$remote_name` is a directory  `$flatpak_ref` contains the app ID and kind (e.g. `app/org.gnome.evince/x86_64/master`).

Handling each app as a separate file also means the file can be deleted after each app is installed, which makes recovery from a later error a little easier.

(Note: ‘repo’ and ‘remote’ mean different things. From the context here, it looks like the code is actually dealing with a remote, rather than a repo. A repo is an OSTree repository. A remote is a local repo’s view of a non-local repo, which necessarily includes some configuration in the local repo’s `config` file.)